Story viewer signup flow

kevintab95 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. ~This PR fixes or fixes part of #[fill_in_number_here].~
2. This PR does the following:

1. This PR introduces sign-in / sign-up flow for logged out users who want to carry on their progress after login.
Screenshots:
![chap](https://user-images.githubusercontent.com/11008603/98310017-fef54b80-1ff1-11eb-8859-99bf6a79db49.png)
![exp](https://user-images.githubusercontent.com/11008603/98310020-00bf0f00-1ff2-11eb-91c6-4df8e3582ac0.png)


- Added lock icons + blurring of locked chapters.
![lock](https://user-images.githubusercontent.com/11008603/98426990-3ed73400-20c1-11eb-9419-21833495f7a8.gif)

- Update classroom promos text.


- Add a loading component to skill selector component. Fixes part of #10822.
![loading_skill_selector](https://user-images.githubusercontent.com/11008603/98471472-83beb000-2212-11eb-9ee7-618bbc60e8b6.gif)


## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

Commit:
Introduce signup flow in the story viewer

Commit:
minor lint issues

Commit:
add suite to check_e2e_tests_are_captured_in_ci

vinitamurthi reviewed (approved):
LGTM from codeowners perspective

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -128,6 +120,69 @@ def post(self, story_id, node_id):
+        story = story_fetchers.get_story_by_id(story_id)
+        completed_nodes = story_fetchers.get_completed_nodes_in_story(
+            self.user_id, story_id)
+        ordered_nodes = [
+            node for node in story.story_contents.get_ordered_nodes()
+        ]
```
Why not?
```suggestion
        ordered_nodes = story.story_contents.get_ordered_nodes()
```

kevintab95 replied:
Oops, fixed.

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -102,16 +104,6 @@ def post(self, story_id, node_id):
-
         next_exp_ids = []
         next_node_id = None
         if not node_id in completed_node_ids:
```
```suggestion
        if node_id not in completed_node_ids:
```

kevintab95 replied:
done.

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -128,6 +120,69 @@ def post(self, story_id, node_id):
+        completed_node_ids = [
+            completed_node.id for completed_node in completed_nodes]
+
+        ordered_nodes = [
+            node for node in story.story_contents.get_ordered_nodes()
+        ]
```
ditto
```suggestion
        ordered_nodes = story.story_contents.get_ordered_nodes()
```

kevintab95 replied:
Done.

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -128,6 +120,69 @@ def post(self, story_id, node_id):
+                    story_url_fragment))
+            return
+
+        (next_exp_ids, next_node_id, _) = (
+            self._record_node_completion(story_id, node_id, [], ordered_nodes))
```
In this place, the `next_node_id` can only be the first node right? Because anyone with completed nodes is redirected and if the node_id is not the first node ID then the user is also redirected.

kevintab95 replied:
`next_node_id` is expected to be the 2nd node, if it exists. If the story does not have a second node, it will return None and in that case we redirect to the story page.

vojtechjelinek replied:
Ah, I get it now, do we actualy record any completion here (because the function is named `_record_node_completion`) might be better to use some different name.

kevintab95 replied:
We actually do record completion of the node in the story. See this:

https://github.com/oppia/oppia/blob/fca6f514a75a5422ce7e0009f51e218dd5d57655/core/controllers/story_viewer.py#L110-L111

vojtechjelinek replied:
Ah, I see.

vojtechjelinek commented on core/tests/protractor_utils/users.js:
```diff
@@ -79,6 +79,42 @@ var _completeSignup = async function(username) {
   await waitFor.pageToFullyLoad();
 };
 
+var completeLoginFlowFromStoryViewerPage = async function(email, username) {
+  var driver = browser.driver;
+  var loginButton = element(by.css('.protractor-test-login-button'));
+  await waitFor.elementToBeClickable(
+    loginButton,
+    'Login button takes too long to be clickable.');
+  await loginButton.click();
+
+  // The statement below uses a browser.wait() to determine if the user has
+  // logged in. Use of waitFor is not possible because the active page is
+  // non-angular.
+  await browser.wait(
+    async() => {
+      try {
+        await driver.findElement(protractor.By.name('email'));
+      } catch (error) {
+        return false;
+      }
+      return true;
+    }, waitFor.DEFAULT_WAIT_TIME_MSECS, 'Login takes too long.');
+  await (await driver.findElement(protractor.By.name('email'))).clear();
+  await (await driver.findElement(protractor.By.name('email'))).sendKeys(email);
+  await (await driver.findElement(protractor.By.id('submit-login'))).click();
+  await waitFor.pageToFullyLoad();
+  var usernameInput = element(by.css('.protractor-test-username-input'));
+  var agreeToTermsCheckbox = element(
+    by.css('.protractor-test-agree-to-terms-checkbox'));
+  var registerUser = element(by.css('.protractor-test-register-user'));
+  await waitFor.visibilityOf(
+    usernameInput, 'No username input field was displayed');
+  await usernameInput.sendKeys(username);
+  await agreeToTermsCheckbox.click();
+  await registerUser.click();
+  await waitFor.pageToFullyLoad();
+};
```
Would it be possible to integrate this with the already existing login/signup functions in this file?

kevintab95 replied:
Yup, done.

vojtechjelinek commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,159 @@
+        true
+      );
+      var url = await browser.getCurrentUrl();
+      dummyExplorationIds.push(url.split('/')[4].replace('#', ''));
```
Isn't there a simple way to get the explorations ID?

kevintab95 replied:
Yep, I noticed we have a `general.getExplorationIdFromEditor()`. I've used that instead now.

vojtechjelinek commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,159 @@
+      'math', 'topic-spf-one', 'story-player-spf-one');
+    await topicAndStoryViewerPage.expectLockedChaptersToBe(2);
+    await topicAndStoryViewerPage.goToChapterIndex(0);
+    await explorationPlayerPage.submitAnswer('Continue', null);
```
Should you check below this that the log in and sign up bettons are visible?

vojtechjelinek replied:
Oh, I see you do it in the `completeLoginFlowFromStoryViewerPage` it would probably make sense to check it here and then handle the login in the users.

kevintab95 replied:
I've moved this to `TopicAndStoryViewerPage.js`.

vojtechjelinek commented on core/templates/services/user.service.ts:
```diff
@@ -76,9 +77,12 @@ angular.module('oppia').factory('UserService', [
           data: newProfileImageDataUrl
         });
       },
+      setReturnUrl: function(newReturnUrl) {
+        returnUrl = newReturnUrl;
+      },
```
Why not `getLoginUrlAsync(returnUrl: string | undefined)` then you don't need to have a service variable.

kevintab95 replied:
Apart from the new login flow from the story / lesson pages, `getLoginUrlAsync` is also used in the `top-navigation-bar.directive` which is present for all pages, so I didn't want to add this as a special case in that directive. Though I'm open to discussing this.

vojtechjelinek replied:
When the `newReturnUrl` is allowed as `undefined` it would be fine to call `getLoginUrlAsync()` so no need to change it elsewhere and it makes sense to me to do it this way instead of this.

kevintab95 replied:
Just to check, do you suggest adding logic in `top-navigation-bar.directive` to detect if the current page is the exploration page and the player is at the last card, then pass in `newReturnUrl`?

I had considered this, and felt that it may be cleaner to have these checks in a player specific directive instead of a general one.

vojtechjelinek replied:
Ah I see, we introduce new buttons that are used to sign in but also we want to modify the behaviour of the buttons in the topnavbar so that they preserve also redirect to the correct places after the login, in that case, this approach makes sense.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

Commit:
address comments; Add blur + lock icon for locked chapters

Commit:
fix lint

Commit:
Merge remote-tracking branch 'upstream/develop' into story-viewer-signup-flow

kevintab95 reviewed (commented):
I've addressed your comments @vojtechjelinek, PTAL!

U8NWXD commented on core/tests/protractor_utils/workflow.js:
```diff
@@ -178,7 +178,7 @@ var createAddExpDetailsAndPublishExp = async function(
 // Creates and publishes a exploration with two cards.
 var createAndPublishTwoCardExploration = async function(
-    title, category, objective, language) {
+    title, category, objective, language, correctnessFeedbackIsEnabled) {
```
This parameter controls whether we enable correctness feedback, so maybe we should name it `enableCorrectnessFeedback` instead? Right now it sounds like it should describe whether correctness feedback has already been enabled

kevintab95 replied:
I think the preferred pattern for boolean variables is this way (see [style guidelines](https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript)). `enableCorrectnessFeedback` sounds more like a function name because it starts with a verb (variables need should not start with a verb, per style guide).

U8NWXD commented on core/tests/protractor_utils/TopicAndStoryViewerPage.js:
```diff
@@ -0,0 +1,53 @@
+    await waitFor.pageToFullyLoad();
+  };
+
+  this.goToChapterIndex = async function(index) {
+    var chapter = await chapterTitleList.get(index);
+    await action.click('Chapter title', chapter);
+    await waitFor.pageToFullyLoad();
+  };
```
Any chance we could identify chapters by name instead?

kevintab95 replied:
I think index is preferred in this case because the order of chapters actually matter and we want to identify failures if this is not the case.

U8NWXD commented on core/tests/protractor_utils/TopicAndStoryViewerPage.js:
```diff
@@ -0,0 +1,53 @@
+    await waitFor.pageToFullyLoad();
+  };
+
+  this.expectLockedChaptersToBe = async function(value) {
```
Naming suggestions:

```suggestion
  this.expectNumLockedChaptersToBe = async function(count) {
```

kevintab95 replied:
I've renamed it to `expectLockedChaptersCountToBe`.

U8NWXD commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,158 @@
+    };
+    for (var i = 1; i <= 3; i++) {
+      await workflow.createAndPublishTwoCardExploration(
+        `Exploration SPF1 - ${i}`,
```
What does the SPF1 mean? Also, can we use `topicAndStoryViewer` in the exploration name for uniqueness?

kevintab95 replied:
The `describe` block was previously called "Story Player Functionality" so I used SPF to refer to stories / topics in this block. I have change it to `topicAndStoryViewer` now.

Done.

U8NWXD commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,158 @@
+
+  it('should save story progress on login.', async function() {
+    await users.createAndLoginAdminUser(
+      'creator@storyPlayer.com', 'creatorStoryPlayer');
```
How about `storyViewer` instead to match the suite name?

kevintab95 replied:
Done.

U8NWXD commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,158 @@
+    var handle = await browser.getWindowHandle();
+    await topicsAndSkillsDashboardPage.get();
+    await topicsAndSkillsDashboardPage.createTopic(
+      'Topic SPF1', 'topic-spf-one', 'Description', false);
```
ditto: Why SPF?

kevintab95 replied:
done.

U8NWXD commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,158 @@
+    (
+      await
+      topicsAndSkillsDashboardPage.createSkillWithDescriptionAndExplanation(
+        'Skill SPF1', 'Concept card explanation', false));
```
ditto: Why SPF1?

kevintab95 replied:
done.

U8NWXD commented on core/tests/protractor_desktop/topicAndStoryViewer.js:
```diff
@@ -0,0 +1,158 @@
+    await topicsAndSkillsDashboardPage.get();
+    await topicsAndSkillsDashboardPage.editTopic('Topic SPF1');
+    await topicEditorPage.addSubtopic(
+      'Subtopic SPF1', 'subtopic-spf-one', Constants.TEST_SVG_PATH,
```
ditto: Why SPF1?

kevintab95 replied:
done.

vojtechjelinek commented on core/tests/protractor_utils/users.js:
```diff
@@ -23,11 +23,26 @@ var waitFor = require('./waitFor.js');
 var adminPage = new AdminPage.AdminPage();
 
-var login = async function(email, isSuperAdmin = false) {
+var login = async function(
+    email, isSuperAdmin = false, manualNavigation = true) {
```
Could you please add an explanation for the `manualNavigation` arg in some comment? (I understand it but without context, it might be hard to guess why is this exactly needed.)

kevintab95 replied:
Done.

vojtechjelinek commented on core/tests/protractor_utils/users.js:
```diff
@@ -58,14 +75,16 @@ var logout = async function() {
 // The user needs to log in immediately before this method is called. Note
 // that this will fail if the user already has a username.
-var _completeSignup = async function(username) {
+var _completeSignup = async function(username, manualNavigation = true) {
```
ditto, add an explanation for `manualNavigation`

kevintab95 replied:
Done.

vojtechjelinek reviewed (changes requested):
Thanks, few more comments.

DubeySandeep reviewed (approved):
LGTM for the codeowner file : scripts/check_e2e_tests_are_captured_in_ci.py

aks681 commented on core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.ts:
```diff
@@ -1238,6 +1252,14 @@ angular.module('oppia').directive('conversationSkin', [
             CurrentInteractionService.submitAnswer();
           };
 
+          $scope.signIn = function() {
```
So, this occurs after the first chapter is completed, right? So, is the user's first chapter progress autosaved?

kevintab95 replied:
Yep! On successful login, the chapter progress will be saved and then the user is redirected to the next chapter.

https://github.com/oppia/oppia/blob/fca6f514a75a5422ce7e0009f51e218dd5d57655/core/controllers/story_viewer.py#L148-L168

Commit:
Minor changes:
- Address review comments.
- Update classroom promos text.
- Add loading component to skill selector component.

kevintab95 reviewed (commented):
This PR is ready for another review pass. PTAL @vojtechjelinek @U8NWXD @aks681!

vojtechjelinek commented on core/tests/protractor_utils/users.js:
```diff
@@ -23,11 +23,29 @@ var waitFor = require('./waitFor.js');
   var driver = browser.driver;
-  await driver.get(general.SERVER_URL_PREFIX + general.LOGIN_URL_SUFFIX);
+  // The manualNavigation argument is used to determine whether to navigate to
+  // the login URL using browser.get() or not. If false, the calling method
```
```suggestion
  // the login URL using driver.get() or not. If false, the calling method
```
or
```suggestion
  // the login URL using browser.driver.get() or not. If false, the calling method
```

kevintab95 replied:
Done.

vojtechjelinek reviewed (approved):
LGTM! Just one minor nit.

Commit:
update to a comment

kevintab95 commented:
@aks681 @U8NWXD please prioritise reviewing this PR. It is currently blocking the release cut.

kevintab95 commented:
Reruns:
![image](https://user-images.githubusercontent.com/11008603/98590970-94942200-22f5-11eb-96c8-002612a49a82.png)
![image](https://user-images.githubusercontent.com/11008603/98591017-a5dd2e80-22f5-11eb-8566-2321d3ebe946.png)

Failures were not related to the new suite.

aks681 reviewed (approved):
Lgtm!

U8NWXD reviewed (approved):
lgtm
